### PR TITLE
WIP: Apply Tuple allocation elimination

### PR DESCRIPTION
**Pull request by forki**
references #330 - Let's see if this still works.

Original idea by @zeux from https://gist.github.com/zeux/1638442

### CODE REVIEW

**Comment by forki** [comment]
In ed264fb I added two small samples.

dotpeek gives the following:

    // Type: FSharp.Compiler.Unittests.TupleEliminationTest
    // Assembly: FSharp.Compiler.Unittests, Version=4.4.0.9055, Culture=neutral, PublicKeyToken=f536804aa0eb945b
    // MVID: 5512E8B3-D40A-62BC-A745-0383B3E81255
    // Assembly location: D:\code\fsharp\Debug\net40\bin\FSharp.Compiler.Unittests.dll
    
    using \u003CStartupCode\u0024FSharp\u002DCompiler\u002DUnittests\u003E;
    using Microsoft.FSharp.Core;
    using NUnit.Framework;
    using System;
    using System.Collections.Generic;
    
    namespace FSharp.Compiler.Unittests
    {
      [TestFixture]
      [CompilationMapping(SourceConstructFlags.ObjectType)]
      [Serializable]
      public class TupleEliminationTest
      {
        public TupleEliminationTest()
        {
          TupleEliminationTest tupleEliminationTest = this;
        }
    
        [Test]
        public void RunLookup1()
        {
          Assert.AreEqual((object) "", (object) FSharpFunc<Dictionary<int, string>, int>.InvokeFast<string>((FSharpFunc<Dictionary<int, string>, FSharpFunc<int, string>>) new \u0024TupleAllocationElimination.RunLookup1\u004023\u002D1((FSharpFunc<IDictionary<int, string>, FSharpFunc<int, string>>) new \u0024TupleAllocationElimination.RunLookup1\u004023(this)), new Dictionary<int, string>(), 42));
        }
    
        [CompilationArgumentCounts(new int[] {1, 1})]
        public static string LookUp(IDictionary<int, string> d, int key)
        {
          string str1 = (string) null;
          Tuple<bool, string> tuple = new Tuple<bool, string>(d.TryGetValue(key, out str1), str1);
          string str2 = tuple.Item2;
          if (tuple.Item1)
            return str2;
          else
            return "";
        }
    
        [Test]
        public void RunLookup2()
        {
          Assert.AreEqual((object) "", (object) TupleEliminationTest.LookUp((IDictionary<int, string>) new Dictionary<int, string>(), 42));
        }
      }
    }

so the function `lookup` gets inlined and no tuple is allocated.
Unfortunately this doesn't work with the static member.

**Comment by KevinRansom** [comment]
Can we find the original author and have him submit this?  We need to be careful about the provenance of pull requests.

Thank you

Kevin

**Comment by zeux** [comment]
I'm the original author; I forgot about this patch when it was rejected on MS Connect, glad to see F# development being more open now!

I'm not 100% sure why it would not work in a static member...

**Comment by forki** [comment]
I'm just trying to run the builds and see if it actually works. 

With 9d4f163 seems I don't have success. dotpeek still shows:

    // Type: FSharp.Compiler.Unittests.TupleElimination
    // Assembly: FSharp.Compiler.Unittests, Version=4.4.0.9055, Culture=neutral, PublicKeyToken=f536804aa0eb945b
    // MVID: 5512F62B-D40A-62BC-A745-03832BF61255
    // Assembly location: D:\code\fsharp\Debug\net40\bin\FSharp.Compiler.Unittests.dll
    
    using Microsoft.FSharp.Core;
    using System;
    using System.Collections.Generic;
    
    namespace FSharp.Compiler.Unittests
    {
      [CompilationMapping(SourceConstructFlags.Module)]
      public static class TupleElimination
      {
        [CompilationArgumentCounts(new int[] {1, 1})]
        public static string lookUp(IDictionary<int, string> d, int key)
        {
          string str1 = (string) null;
          Tuple<bool, string> tuple = new Tuple<bool, string>(d.TryGetValue(key, out str1), str1);
          string str2 = tuple.Item2;
          if (tuple.Item1)
            return str2;
          else
            return "";
        }
      }
    }

**Comment by forki** [comment]
@zeux: seems I don't get it to work at all.

**Comment by zeux** [comment]
Note: the change essentially tried to transform the code that was generated for automatic out-value to tuple conversion into code that is optimized by tuple allocation elimination.

If either the conversion code changed to generate a different structure or allocation code changed this may no longer be effective and has to be adjusted...

**Comment by forki** [comment]
@asik: in #330 you said you tested the change. Could you please describe how.

**Comment by asik** [comment]
I compiled a little sample like you did in release mode and looked at the dissassembly in ILDasm. That was a few months ago though. I can try again later today to see if that still works.

**Comment by forki** [comment]
Yes. Please try with my sample. Maybe I'm just holding it wrong.

**Comment by asik** [comment]
If you're not using --optimize+ then that code isn't invoked at all.

**Comment by mrange** [comment]
I think this is a small but important PR that allows us to write idiomatic code with no overhead.

If there's an interest I can setup some IL tests to verify that the tuple is actually eliminated

**Comment by forki** [comment]
Yes. I'd love to see that.

File: src/fsharp/opt.fs [hunk 1]
```diff
@@ -1440,6 +1442,34 @@ let ExpandStructuralBinding cenv expr =
...
+// let t = (let a0=v0 in let a1=v1 in ... in let an=vn in e0,e1,...,em) in body
+// let a0=v0 in let a1=v1 in ... in let an=vn in (let t = e0,e1,...,em in body)
+// This way ExpandStructuralBinding can replace expressions in constants, t is directly bound
+// to a tuple expression so that other optimizations such as OptimizeTupleFieldGet work,
+// and the tuple allocation can be eliminated.
+// Most importantly, this successfully eliminates tuple allocations for implicitly returned
+// formal arguments in method calls.
+let rec RearrangeTupleBindings expr fin =
+    match expr with
+    | Expr.Let (bind,body,m,_) ->
```
Comment (dsyme) [inline src/fsharp/opt.fs]: This optimization is generally useful and can be expanded to include Expr.Sequential and Expr.LetRec

For Expr.Sequential:

    let t = 
        let a = v1
        stmt-expr
        let b = v2
        a,b

File: src/fsharp/opt.fs [hunk 2]
```diff
@@ -1440,6 +1442,34 @@ let ExpandStructuralBinding cenv expr =
...
+        match RearrangeTupleBindings body fin with
+        | Some b -> Some (mkLetBind m bind b)
+        | None -> None
+    | Expr.Op (TOp.Tuple,_,_,_) ->
```
Comment (dsyme) [inline src/fsharp/opt.fs]: In principle this optimization can flatten a much broader range of expressions, e.g. ones that end in any TOp, App, Lambda etc.

I don't think I can immediately convince you that we should do this, but the nesting-of-let-bindings is common in the compiler, and we should consider over time whether we extend the applicability of this technique.

**Comment by forki** [comment]
@zeux did you already sign the CLA and are you OK with this contribution? We use this here for discussion and you could send your own PR.

**Comment by mrange** [comment]
@forki I created a simple IL test for to verify elimination of tuples. You can find it here: https://github.com/mrange/visualfsharp/commit/7140a0e77082533aeeff459b977850673304a833

File: src/fsharp/opt.fs (hunk 2 above)
Comment (asik) [inline src/fsharp/opt.fs]: If it results in faster code, please do!

**Comment by zeux** [comment]
I did not sign the CLA; I'm happy to sign it and to make a PR!

**Comment by forki** [comment]
cool.

btw: as expected in release mode both gets optimized:

    // Type: FSharp.Compiler.Unittests.TupleElimination
    // Assembly: FSharp.Compiler.Unittests, Version=4.4.0.9055, Culture=neutral, PublicKeyToken=f536804aa0eb945b
    // MVID: 551313B5-D40A-62BC-A745-0383B5131355
    // Assembly location: D:\code\fsharp\Release\net40\bin\FSharp.Compiler.Unittests.dll
    
    using Microsoft.FSharp.Core;
    using System.Collections.Generic;
    
    namespace FSharp.Compiler.Unittests
    {
      [CompilationMapping(SourceConstructFlags.Module)]
      public static class TupleElimination
      {
        [CompilationArgumentCounts(new int[] {1, 1})]
        public static string lookUp(IDictionary<int, string> d, int key)
        {
          string str1 = (string) null;
          bool flag = d.TryGetValue(key, out str1);
          string str2 = str1;
          if (flag)
            return str2;
          else
            return "";
        }
      }
    }


    // Type: FSharp.Compiler.Unittests.TupleEliminationTest
    // Assembly: FSharp.Compiler.Unittests, Version=4.4.0.9055, Culture=neutral, PublicKeyToken=f536804aa0eb945b
    // MVID: 551313B5-D40A-62BC-A745-0383B5131355
    // Assembly location: D:\code\fsharp\Release\net40\bin\FSharp.Compiler.Unittests.dll
    
    using Microsoft.FSharp.Core;
    using NUnit.Framework;
    using System;
    using System.Collections.Generic;
    
    namespace FSharp.Compiler.Unittests
    {
      [TestFixture]
      [CompilationMapping(SourceConstructFlags.ObjectType)]
      [Serializable]
      public class TupleEliminationTest
      {
        public TupleEliminationTest()
        {
          TupleEliminationTest tupleEliminationTest = this;
        }
    
        [Test]
        public void RunLookup1()
        {
          Assert.AreEqual((object) "", (object) TupleElimination.lookUp((IDictionary<int, string>) new Dictionary<int, string>(), 42));
        }
    
        [CompilationArgumentCounts(new int[] {1, 1})]
        public static string LookUp(IDictionary<int, string> d, int key)
        {
          string str1 = (string) null;
          bool flag = d.TryGetValue(key, out str1);
          string str2 = str1;
          if (flag)
            return str2;
          else
            return "";
        }
    
        [Test]
        public void RunLookup2()
        {
          Assert.AreEqual((object) "", (object) TupleEliminationTest.LookUp((IDictionary<int, string>) new Dictionary<int, string>(), 42));
        }
      }
    }


Good work. I will now try to add some bigger perf tests in order to meet the contributor guidelines for  perf optimizations.

**Comment by forki** [comment]
I still wonder why we need to introduce `str2`

**Comment by mrange** [comment]
A thought is that perf tests should probably include gc time. the benefit might be small on platforms with a competent gc. however for game devs on platforms with poor gcs this could give good improvements. iirc the example posted by @zeux seemed to be game related.

**Comment by mrange** [comment]
str2 might well be eliminated by the jitter. can check tomorrow....

**Comment by zeux** [comment]
As far as performance goes, what sort of benchmark is expected?

It's simple to set up something trivial that stresses the allocation - for example, this:

```F#
open System.Collections.Generic
open System.Diagnostics
open System

let cache = Dictionary<int, int64>()

let rec square n =
    match cache.TryGetValue(n) with
    | true, v -> v
    | _ ->
        let v =
            if n < 1 then
                0L
            else
                square (n - 1) + (int64 n * 2L - 1L)
        cache.Add(n, v) |> ignore
        v

let sw = Stopwatch.StartNew()

let mutable sum = 0L
for i=1 to 100 do
    cache.Clear()
    let res = square 10000
    sum <- sum + res

let time = sw.Elapsed

printfn "%d in %A" sum time
```

Gives the following results:

Baseline:
10000000000 in 00:00:00.0493286

With patch:
10000000000 in 00:00:00.0378752

**Comment by forki** [comment]
@mrange I applied your test in 8067a90 (without applying tuple elimination) but build is not red.

**Comment by forki** [comment]
@zeux 

For the [Contributing GuideLines](https://github.com/Microsoft/visualfsharp/blob/fsharp4/CONTRIBUTING.md):

* Performance tests and figures must be given, either in the PR or in the notes associated with the PR.
* The PR must show a reliable, substantive performance improvement that justifies the complexity introduced. For the compiler, performance improvements of ~1% are of interest.

With 9a0ea5e I can reproduce your benchmark results, which are much better than 1% for this case. I think this criteria is met.

**Comment by dsyme** [comment]
Strictly speaking the 1% refers to the perf of the compiler itself. A bit more perf testing over longer runs and using #time in F# Interactive to get GC figures would be useful. Also adfing some IL codegen baseline tests would be great.

**Comment by forki** [comment]
> Strictly speaking the 1% refers to the perf of the compiler itself.

ahh ;-)

**Comment by zeux** [comment]
More iterations and fsi output: https://gist.github.com/zeux/12fb2e40994b8c142960

A lot of GC collections without the patch and no collections with :)

**Comment by latkin** [comment]
full of win

**Comment by zeux** [comment]
FWIW I've verified locally that reverting the change to opt.fs makes the test fail. It took me some time since there was an issue with a hosted compiler - it used the code with the change to opt.fs even after I reverted the change - but this is probably because I used Visual Studio to build fsc.exe. I verified the tests by disabling hosted compiler.

I'm still working on getting CLA signed; in the meantime I'll submit a PR with a change to opt.fs and @mrange's test.

**Comment by forki** [comment]
> it used the code with the change to opt.fs even after I reverted the change

I assume you have to rebuild the proto-compiler. I use the following:
* `git clean -xdf`
* `appveyor-build.cmd`

**Comment by forki** [comment]
closing this in favour of https://github.com/Microsoft/visualfsharp/pull/335

**Comment by latkin** [comment]
> there was an issue with a hosted compiler - it used the code with the change to opt.fs even after I reverted the change 

After making compiler changes you need to make sure to re-run `update.cmd` to update the GAC, that's where the test tools will be grabbing FSharp.Core from.  Depending on the change you might need to reorun `BuildTestTools.cmd` as well.

**Pull request closed without merging**